errors/errors.go

```diff
@@ -0,0 +1,212 @@
+//   - error reason
+//   - metadata information
+//   - optional resourceInfo (componenttype/name)
+type Error struct {
+	err            error
+	description    string
+	reason         string
+	httpCode       int
+	grpcStatusCode codes.Code
+	metadata       map[string]string
+	resourceInfo   *ResourceInfo
+}
+
+// New create a new Error using the supplied metadata and Options
+func New(err error, metadata map[string]string, options ...Option) *Error {
+	if err == nil {
+		return nil
+	}
+
+	// Use default values
+	de := &Error{
+		err:            err,
+		reason:         unknown,
+		httpCode:       unknownHTTPCode,
+		grpcStatusCode: codes.Unknown,
+	}
+
+	// Now apply any requested options
+	// to override
+	for _, option := range options {
+		option(de)
+	}
+
+	return de
+}
+
+// Error implements the error interface.
+func (e *Error) Error() string {
+	if e != nil && e.err != nil {
+		return e.err.Error()
+	}
+	return ""
+}
+
+// Unwrap implements the error unwrapping interface.
+func (e *Error) Unwrap() error {
+	return e.err
+}
+
+// Description returns the description of the error.
+func (e *Error) Description() string {
+	if e.description != "" {
+		return e.description
+	}
+	return e.err.Error()
+}
+
+// WithErrorReason used to pass reason, httpCode, and
+// grpcStatus code to the Error struct.
+func WithErrorReason(reason string, httpCode int, grpcStatusCode codes.Code) Option {
+	return func(err *Error) {
+		err.reason = reason
+		err.grpcStatusCode = grpcStatusCode
+		err.httpCode = httpCode
+	}
+}
+
+// WithResourceInfo used to pass ResourceInfo to the Error struct.
+func WithResourceInfo(resourceInfo *ResourceInfo) Option {
+	return func(e *Error) {
+		e.resourceInfo = resourceInfo
+	}
+}
+
+// WithDescription used to pass a description
+// to the Error struct.
+func WithDescription(description string) Option {
+	return func(e *Error) {
+		e.description = description
+	}
+}
+
+// WithMetadata used to pass a Metadata[string]string
+// to the Error struct.
+func WithMetadata(md map[string]string) Option {
+	return func(e *Error) {
+		e.metadata = md
+	}
+}
+
+func newErrorInfo(reason string, md map[string]string) *errdetails.ErrorInfo {
+	return &errdetails.ErrorInfo{
+		Domain:   Domain,
+		Reason:   reason,
+		Metadata: md,
+	}
+}
+
+func newResourceInfo(rid *ResourceInfo, err error) *errdetails.ResourceInfo {
+	return &errdetails.ResourceInfo{
+		ResourceType: rid.Type,
+		ResourceName: rid.Name,
+		Owner:        Owner,
+		Description:  err.Error(),
+	}
+}
+
+// *** GRPC Methods ***
+
+// GRPCStatus returns the gRPC status.Status object.
+func (e *Error) GRPCStatus() *status.Status {
+	messages := []protoiface.MessageV1{
+		newErrorInfo(e.reason, e.metadata),
+	}
+
+	if e.resourceInfo != nil {
+		messages = append(messages, newResourceInfo(e.resourceInfo, e.err))
+	}
+
+	ste, stErr := status.New(e.grpcStatusCode, e.description).WithDetails(messages...)
+	if stErr != nil {
+		return status.New(e.grpcStatusCode, e.description)
```

Comment:
Shouldn't this return an Internal Server Error? If I understand correctly, `WithDetails` returns an error if it fails to parse messages, which should indicate a development-time issue and not a runtime one.


```suggestion
		return status.New(codes.Internal, fmt.Sprintf("failed to create gRPC status message: %v", stErr))
```